### PR TITLE
TST/DOC: Set `jax_default_matmul_precision` to `float32` for tests using the JAX backend

### PR DESCRIPTION
**Pull request by steppi**
#### What does this implement/fix?

In @lucascolley's PR https://github.com/scipy/scipy/pull/24714, some tests for `LinearOperator` were showing very large relative error on order of `1e-2` when using the JAX backend on GPU. I determined that this is because JAX defaults to using [TensorFloat32](https://en.wikipedia.org/wiki/TensorFloat-32) when multiplying `float32` arrays on GPUs that support this floating point format, as described here: https://github.com/jax-ml/jax/issues/4873.

Though very useful for neural networks, the half precision mantissas of the `TensorFloat32` format can cause serious accuracy issues in scientific applications, so I think it's a good idea to set `jax_default_matmul_precision="float32"` in `conftest.py` and document this is the Array API docs. I've done both of these things here.

#### Additional information

There is a test in `scipy.differentiate.tests.test_differentiate` which passes before this option has changed and fails afterwards. The failure is due to a 1 ulp discrepancy in the value of a vectorized `jax.numpy.exp` vs separate applications of the scalar `jax.numpy.exp` on GPU. I don't think there is a good reason why adjusting the `jax_default_matmul_precision` would cause this discrepancy. It seems to be due to some strange quirk in code generated by the JAX compiler. I simply adjusted the tolerance.

#### AI Generation Disclosure
This PR contains no AI generated code or text. I did however ask Google Gemini  "why might JAX be less accurate for float32 matrix multiplications on GPU than PyTorch or CuPy?" and it was what suggested default `TensorFloat32` precision as a potential underlying reason. I did not know this fact off-hand. In the pre-AI era I likely would have found this information through a traditional search engine query.

### CODE REVIEW

**Review by lucascolley — APPROVED** [review]
thanks Albert, LGTM! But let's check the rendered docs before merge.

**Comment by steppi** [comment]
Thanks @lucascolley!

**Comment by steppi** [comment]
This is hilarious to me:

image

Will fix and push a `[docs only]` commit.

**Comment by steppi** [comment]
I took the time to build the docs locally this time. This looks much better:

image
